src/printer.cpp

```diff
@@ -0,0 +1,129 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright (C) 2020 YADRO
+
+#include "printer.hpp"
+
+#include <json.h>
+
+void Printer::setNameFilter(const char* name)
+{
+    nameFilter_ = name;
+}
+
+void Printer::allowNonexitent()
+{
+    allowNonexitent_ = true;
+}
+
+void Printer::allowEmptyProperties()
+{
+    allowEmptyProperties_ = true;
+}
+
+void Printer::printText(const std::vector<InventoryItem>& items) const
+{
+    // Size of the column with property name (formatting output)
+    static const size_t PropNmColWidth = 20;
+
+    for (const InventoryItem& item : items)
+    {
+        // filter out by name
+        if (!nameFilter_.empty() && nameFilter_ != item.name)
+            continue;
+
+        // filter out nonexistent items
+        if (!allowNonexitent_ && !item.isPresent())
+            continue;
+
+        // print title
+        printf("%s: %s\n", item.name.c_str(), item.prettyName().c_str());
+
+        // print properties
+        for (const auto& property : item.properties)
+        {
+            // get value from variant
+            std::string val;
+            std::visit(
+                [&val](auto&& arg) {
+                    using T = std::decay_t<decltype(arg)>;
+                    if constexpr (std::is_same_v<T, bool>)
+                        val = arg ? "Yes" : "No";
+                    else if constexpr (std::is_arithmetic<T>::value)
+                        val = std::to_string(arg);
+                    else if constexpr (std::is_same_v<T, std::string>)
+                        val = arg;
+                    else
+                        static_assert(T::value, "Unhandled value type");
+                },
+                property.second);
+
+            // filter out empty properties
+            if (!val.empty() || allowEmptyProperties_)
+            {
+                const size_t nameLen = property.first.length();
+                printf("  %s: ", property.first.c_str());
+                if (nameLen < PropNmColWidth)
+                    printf("%*s", PropNmColWidth - nameLen, "");
```

Comment:
I've tried to build it and got:
```
error: field width specifier ‘*’ expects argument of type ‘int’, but argument 2 has type ‘long unsigned int’ [-Werror=format=]
```

Reply:
Besides, I thought that all these printf may be joined to the one call.
```cpp
printf(" %s:%*s%s\n", property.first.c_str(),
       nameLen < PropNmColWidth ? (int)(PropNmColWidth - nameLen) : 0, "",
       val.c_str());
```

Reply:
It's strange, but I don't see this error. Do you use master branch?
Anyway, fixed.

Reply:
I believe it depends by compiler.
I've tried:
 * C++ compiler for the host machine: c++ (gcc 9.3.1 "c++ (GCC) 9.3.1 20200408 (Red Hat 9.3.1-2)")
 * C++ compiler for the host machine: clang++ (clang 9.0.1 "clang version 9.0.1 (Fedora 9.0.1-2.fc31)")

And both of them stop the build with this error.